rs/replicated_state/src/canister_state/queues.rs

```diff
@@ -1052,7 +1145,7 @@ impl CanisterQueues {
         // same `time_out_messages()` call; `on_message_dropped(1)` will also pop `2`).
         if queue.peek() == Some(reference) {
             queue.pop();
-            queue.pop_while(|reference| self.pool.get(reference).is_none());
+            queue.pop_while(|reference| is_stale(reference, &self.pool, &self.shed_responses));
```

Comment:
Why is it always the right thing to pass `shed_responses` to `is_stale` here? We might be dealing with an output queue no?

Reply:
Technically, we wouldn't pass `shed_responses` if we were dealing with an output queue. But we would need extra logic to distinguish between "request in input queue" (because responses we've already handled above) and "message in output queue". And `is_stale()` explicitly only looks into `shed_responses` if the reference is for an inbound response. So passing `shed_responses` alongside an outbound reference is equivalent to passing an empty map.

Reply:
I've added type parameters (of the type "held" by the queue) to both queues and references in follow-up PR #1697.

With that, this function has split into two (for `InboundReference` vs `OutboundReference` arguments); and this has turned into `self.store.queue_advance(input_queue)` / `self.store.queue_advance(output_queue)` (although the code could, and probably should, be factored out because it's identical excapt for the type of the queue).